Draft for negation types article (#17)
JelleZijlstra/JelleZijlstra.github.io: negation-types.md

spec, which provides a solid foundation for the type system in a way that was previously lacking.
In addition, Astral's new type checker, [ty](https://github.com/astral-sh/ty), is attempting to build
directly on these core concepts and to use more set theory-based concepts in its implementation of the
type system.

These two developments have prompted me to take a closer look at possible extensions of the type system
that align with the set-theoretic framework we now use. In particular, this article covers *negation 
types*: types that contain all values that are not part of some other type.
I'll first review the theoretical framework, then discuss how negation types could fit into Python's
gradual type system, and close by discussing the practical value of negation types.

## A brief tour of the set-theoretic gradual type system

For the following discussion of negation types to make sense, we first have to review the core structure
of the type system, as outlined in more detail in
["Type system concepts"](https://typing.python.org/en/latest/spec/concepts.html).

Python's type system is conceptually based on set-theoretic types: a type is understood as a set of 

> values. For example, the type `int` includes all values that are instance of the class `int` or a 

Reviewer: ```suggestion
possible values. For example, the type `int` includes all values that are instance of the class `int` or a 
```

> subclass at runtime, and the type `Literal[43]` includes only the instance of `int` with the value 

Reviewer: This feels a little imprecise to me: `Literal[42]` must not merely be an instance of int; its `__class__` must be _exactly_ `int`. On the other hand, there may in fact be multiple possible values at runtime that satisfy this type, since `Literal[42]` will not necessarily be a singleton on every implementation of Python; talking about it as "_the_ instance of `int`" implies that there will only ever be one runtime value that satisfies the type, which seems incorrect.

Author: That's what I meant to say but I agree it's not what I am conveying, let me think of a clearer way to write it.

Author: Rewording to "instances of exactly `int` (not a subclass) with the value `43`".

(There is in fact only one instance because of the small int cache, but that's not something the type system should be aware of :) .)

Reviewer: > There is in fact only one instance because of the small int cache

I know, that's why I was careful to say "will not _necessarily_ be a singleton on _every implementation_ of Python" 😉

there's only one instance on CPython, but that won't necessarily be true on PyPy or GraalPy or some other hypothetical implementation of Python — it's a CPython implementation detail rather than something that's part of the language spec

`43`. Subtyping is defined through the subset relation: because all members of `Literal[43]` are also
members of `int`, `Literal[43]` is a subtype of `int` and a value of type `Literal[43]` can be used
when an `int` is expected.

> But what we just said only applies to what we call *fully static types*, which represent a known

Reviewer: ```suggestion
But what we just said only applies to what we call a *fully static type*, which represents a known
```

(or you could edit the latter part of the sentence to "known sets of values")

> set of values. In addition, we have *dynamic types*, which include a *gradual form* such as `Any`.

Reviewer: ```suggestion
set of possible values. In addition, we have *dynamic types*, which include a *gradual form* such as `Any`.
```

> (We might want to use the term "gradual type" for dynamic types, but the spec has chosen to define

Reviewer: I believe the choice here was grounded in the way the term "gradual type" is commonly used in the type-theory literature — it's not a choice unique to the spec

Author: Rewording this to not cast aspersions on the spec

"gradual type" to include all types, both static and dynamic.) A dynamic type does not refer to

> a single set of values, but can refer to any type within a range. We call these types *materializations*,

Reviewer: ```suggestion
a single set of possible values, but can refer to any type within a range. We call these types *materializations*,
```

and in general, operations on gradual types are acceptable if there is any materialization of the gradual
type that could make the operation work. The most prominent gradual type is `Any`, which can materialize to any type.

For example, type checkers emit an error if they see an attempt to add an `int` to a `str` (`int + str`).
But `int + Any` is allowed: the `Any` could materialize to a type that would make the `+` operation work,
such as `int` or `float`. Some errors may still be detected even if `Any` is used, though. Consider the

> operation `int + str | Any`, adding an `int` to something that may be either a `str` or `Any`. Now, no 

Reviewer: This might clarify the order of operations a little:

```suggestion
operation `int + (str | Any)`, adding an `int` to something that may be either a `str` or `Any`. Now, no 
```

`T` (assuming `T` is fully static). For example, the type `~bool` contains all values that are not 
instances of `bool`. In practice we usually encounter negation types in intersections. For example,
the type `int & ~bool` contains all values of type `int` that are not instances of `bool`.

Above I was careful to specify only fully static types. The behavior of negated dynamic types is a
bit less intuitive. For example, `~Any` is in fact the same type as `Any` (in the spec's terminology,
these are equivalent types). That's because `Any` can materialize to any fully static type, and therefore also to the negation of any other fully static type. Thus, for any materialization of `~Any`,
there is an equivalent materialization of `Any`, and vice versa. For example,
if we have the `Any` in `~Any` materialize to `int`, the type is `~int`, and the other `Any` can 
materialize directly to `~int`. If instead we have the first `Any` materialize to `~int`, the type
simplifies to `~~int = int`, and the other `Any` can materialize to `int`.

### Top and bottom materializations

We can generalize this observation to transform away all negated dynamic types, but first we need to

> introduce a new concept: the *top materialization*, or `Top[T]`. For any type `T` (dynamic or static),

Reviewer: ```suggestion
introduce a new concept: the *top materialization*, or `Top[T]`. For any type `T` (dynamic or fully static),
```

Author: Since we don't have "partially static" types as a defined concept (I guess any dynamic type other than Any is "partially static"?), I am subtly nudging towards more concise terminology :)

Reviewer: I guess I'd find it clearer if you either used "static" or "fully static" consistently the whole way through, otherwise it _could_ be taken to imply that there's some kind of distinction being made between two concepts :-)

Author: Fair point, and looks like this was the only place where I didn't use "fully static" so I'll change it.

`Top[T]` is a fully static type. For a fully static type, `Top[T]` is the same as `T`; for a dynamic

> type, it is the union of all materializations, and therefore every materialization is a subtype of

Reviewer: ```suggestion
type, it is the union of all possible materializations, and therefore every possible materialization is a subtype of
```

Author: Why is it important to add "possible" here?

Reviewer: eh, it's probably fine as-is :-)

> `Top[T]`. For example, the type `Top[list[Any]]` includes all materializations of `list[Any]`, and

Reviewer: ```suggestion
`Top[T]`. For example, the type `Top[list[Any]]` includes all possible materializations of `list[Any]`, and
```

therefore all instances of `list` regardless of their generic argument.

We can similarly create a *bottom materialization* `Bottom[T]` that is a subtype of every materialization
of `T`, or the intersection of all materializations of `T`. (We'll need this type later.) These concepts

> appear in the gradual typing literature, but they have some issues when used in the Python type system.

Reviewer: ```suggestion
appear in the gradual typing literature, but they have some issues when applied to the Python type system.
```

First, we call it the "top materialization", but by a literal reading of the spec, this type is not a 
materialization of `list[Any]` at all: the spec
[says](https://typing.python.org/en/latest/spec/concepts.html#materialization) that we create 

> materializations by "replacing" occurrences of `Any` with another type, and we can create a union type

Reviewer: Did you mean to say "but we cannot" rather than "and we can" here?

Author: Yes, rewrote that wrong at some point

> like `list[str] | list[int]` by replacing `Any` with a different type, let alone the infinite union that 

Reviewer: ```suggestion
like `list[str] | list[int]` by replacing `Any` in `list[Any]` with a different type, let alone the infinite union that 
```

Reviewer: I don't think this is necessarily relevant here, but it occurs to me (since I've been working on tuple types recently) that `tuple[int, ...]` is mostly treated like an "infinite Union" of all possible fixed-length tuple types in which all elements are of type `int`

Author: Yes, I think that's an important idea that interacts with the discussion later about attributes on negation types. I think what I landed at is that attributes on negation types are narrowed relative to the non-negation type only if the positive side of the intersection is conceptually a union, and we can throw out some of the union members.

is `Top[list[Any]]`. This is a [known defect](https://github.com/python/typing/issues/2027) in the 
spec; I haven't proposed a fix yet since I need to think more about the right replacement wording,
but a better definition of materialiation would allow for the union or intersection of two 
materializations to also be a materialization.

Second, and more seriously, the bottom materialization of many dynamic types will be equivalent to 
`Never`. An object that is both a `list[int]` and a `list[str]`, for example, cannot exist.
This in turn means that under the spec, all of these dynamic types are assignable to each other,
since they have a shared materialization. This I believe is another defect in the spec, but we need
to think more about how to fix it. For our purposes in this article, we can simply stick our

> head in the sand and pretend that `Never` and `Bottom[T]` are different types. It's not a very satisfying

Reviewer: ```suggestion
heads in the sand and pretend that `Never` and `Bottom[T]` are different types. It's not a very satisfying
```

Unless we all have shared ownership of a single head 😱

solution, but it allows some interesting reasoning. I hope to write more about this topic later.

### Simplifying negation types

Now, with `Top[T]` available as a tool, we can formulate a rule: for a dynamic type `T`, `~T` is
equivalent to `~Top[T] | T`. This is the union of all objects that are not part of any materialization 
of `T`, plus some unknown part of the dynamic type `T`.

We can apply this rule to some example types:

- `~Any` is equivalent to `~Top[Any] | Any`. `Top[Any]` is `object`, so this reduces to
  `~object | Any = Never | Any = Any`.

> - `~tuple[Any] = ~Top[tuple[Any]] | tuple[Any] = ~tuple[object] | tuple[Any]`.

Reviewer: Practically speaking, I believe this type could materialise to any fixed-length tuple type or any type that is a subtype of a fixed-length tuple type, but no type that is a supertype of any tuple type. Is that correct?

E.g. `Sequence[int]` does not satisfy this type: it overlaps with `tuple[object]` so the l.h.s. of the union is not satisfied; and it's not assignable to `tuple[Any]`, so the r.h.s. is also unsatisfied.

But `tuple[str, str]` is disjoint from `tuple[object]`, and therefore a subtype of `~tuple[object]`, therefore a possible materialisation of `~tuple[object] | tuple[Any]`

Author: That's not right; `Sequence[object]` (and anything else) should be assignable to `~tuple[Any]`, since it can materialize to `~tuple[Never] = ~Never = object`. In the expanded version `~tuple[object] | tuple[Any]` we get this by setting the Any to `object` -> `~tuple[object] | tuple[object] = object`.

I think this means that the trick for extracting dynamic types from the negation is not actually going to make life easier for type checkers implementing assignability, since they're going to have to distribute over the union. So to check whether `T` is assignable to `~U`, better to directly check whether `Bottom[T]` and `Bottom[U]` are disjunct.

Reviewer: > `Sequence[object]` (and anything else) should be assignable to `~tuple[Any]`, since it can materialize to `~tuple[Never] = ~Never = object`.

Hmm, but per https://github.com/astral-sh/ty/issues/639#issuecomment-2971958610 I don't think `tuple[Never]` is a valid materialization of `tuple[Any]`, due to the rule

> for `A` to be assignable to `B`, [it is required] that `A` is a consistent subtype of `B`, and that, if the bottom or lower bound materialization of `A` is `Never`, that the upper bound or top materializations of `A` and `B` be non-disjoint. Another way to describe this is that there must be some object that can inhabit some possible materialization of both `A` and `B`, in order for `A` to be assignable to `B`.

There is no object that would both inhabit `Never` and `tuple[Any]`, ergo `Never` is not assignable to `tuple[Any]`, ergo `tuple[Never]` (equivalent to `Never`) is not a valid materialization of `tuple[Any]`, ergo `~tuple[Never]` is not a valid materialization of `~tuple[Any]`.

Author: Exactly how Never interacts with materialization is still up for debate; I think the proposed rule has some questionable consequences. But maybe I can't ignore the materialization/Never issue in this text.

Author: My preferred rule is still that no *top-level* type can materialize to Never, but Anys nested within it can. So for example, Any cannot materialize to Never, but in `~Any`, Any can materialize to Never (but not `object`!) because `~Never = object`. My main hesitation with this rule is that it appears to break the bottom materialization, which is often equivalent to Never.

Reviewer 2: > per [astral-sh/ty#639 (comment)](https://github.com/astral-sh/ty/issues/639#issuecomment-2971958610) I don't think `tuple[Never]` is a valid materialization of `tuple[Any]`

Note that the rule described there (from Elixir) is a rule about _assignability_, not a rule about valid materializations. In other words, it says that for `A` to be assignable to `B`, there must be _non-Never_ materializations `A'` and `B'` for which `A' <: B'`. It doesn't say that `Never` cannot be a materialization of a gradual type.

In practice I think this has similar results to @JelleZijlstra's preferred phrasing. Effectively, "the type we are considering in the context of an assignability relation" gives a more precise meaning to the concept of "top-level type" (which otherwise IMO has no clear meaning, since any given type may be a "top-level type" in some context and not in some other context.)

`Never`. An object that is both a `list[int]` and a `list[str]`, for example, cannot exist.
This in turn means that under the spec, all of these dynamic types are assignable to each other,
since they have a shared materialization. This I believe is another defect in the spec, but we need
to think more about how to fix it. For our purposes in this article, we can simply stick our
heads in the sand and pretend that `Never` and `Bottom[T]` are different types. It's not a very satisfying
solution, but it allows some interesting reasoning. I hope to write more about this topic later.

### Simplifying negation types

Now, with `Top[T]` available as a tool, we can formulate a rule: for a dynamic type `T`, `~T` is
equivalent to `~Top[T] | T`. This is the union of all objects that are not part of any materialization
of `T`, plus some unknown part of the dynamic type `T`.

We can apply this rule to some example types:

- `~Any` is equivalent to `~Top[Any] | Any`. `Top[Any]` is `object`, so this reduces to
  `~object | Any = Never | Any = Any`.
- `~tuple[Any] = ~Top[tuple[Any]] | tuple[Any] = ~tuple[object] | tuple[Any]`.

> Whether this counts as simplifying is debatable, since it usually makes for a more complex type.

Reviewer: Yes, there's often reason to apply a "simplification" that actually results in a more complex type to look at -- we often apply such "complicating simplifications" when constructing unions and intersections in ty: https://github.com/astral-sh/ruff/blob/4bc27133a9c1cecdfd53669662db1292b6c1dc83/crates/ty_python_semantic/src/types/builder.rs#L620-L645

Author: Those mostly look like unarguable simplifications to me :)

It's interesting that you need to write out both directions, e.g. both `AlwaysFalsy & LiteralString = Literal[""]` and `LiteralString & AlwaysFalsy = Literal[""]`. I can see why you'd need to do that but it feels a bit error prone.

Reviewer: It _is_ error-prone; there are still some identities that we don't get right, e.g. https://github.com/astral-sh/ty/issues/216 ☹️ there is a selection of property tests for ty that cannot be stabilised until these issues are fixed, which is quite unfortunate.

We need to fix this at some point, but there are more pressing issues for us right now...

the negative part of the intersection can simply be ignored. A type like
`int & ~Literal[1]` has the same attributes as plain `int`; `object & ~str` does not differ from
`object`.
It is important, of course, that the intersection first gets simplified.
`(int | str) & ~str` does not have the same attributes as `int`.

It gets more complicated with structural types in the negative case. Consider this program:

```python
from typing import Protocol, ReadOnly

class HasIntStr(Protocol):
    x: ReadOnly[int | str]

class HasStr(Protocol):
    x: ReadOnly[str]

obj: HasIntStr & ~HasStr
reveal_type(obj.x)
```

(Assuming that `ReadOnly` is accepted on protocol attributes, as proposed by
[PEP 767](https://peps.python.org/pep-0767).)

Naively, we might think that this should reveal `int`: `obj.x` looks like it should be
`(int | str) & ~str = int`. But there could be implementations of `HasIntStr` that don't
overlap with `HasStr` at all, yet might return either an `int` or a `str` when the attribute
is accessed.

```python
class HasIntStrSneaky:
    @property
    def x(self) -> int | str:
        return 1 if random() > 0.5 else "1"

    @x.setter
    def x(self, obj: int | str) -> None:

>         pass

Reviewer: I think this type _does_ overlap with `HasStr`. I could write a Liskov-compliant subclass of this class like so:

```py
class NothingWrongWithThis(HasIntStrSneaky):
    @HasIntStrSneaky.x.getter
    def x(self) -> str:
        return "foo"
```

Or like this:

```py
class NothingWrongWithThisEither(HasIntStrSneaky):
    @property
    def x(self) -> str:
        return "foo"
    
    @x.setter
    def x(self, obj: int | str) -> None:
        pass
```

This type _is_ a subtype of `HasStr` _and_ a subtype of `HasIntStrSneaky`, meaning that `HasIntStrSneaky` and `HasStr` are not disjoint (they overlap).

- https://mypy-play.net/?mypy=latest&python=3.12&gist=1dd38cb918b689a5cc485f91191d25ab
- https://pyright-play.net/?pythonVersion=3.13&strict=true&enableExperimentalFeatures=true&code=GYJw9gtgBCCGB2ATSUCWEAOYQBcYOQgChQUcBPDVeAczU2zwAVwcwBjMAGyKPa9gBnQVAASQgMo4QAChZg2nLgEoAXESiaoAAQzgMAU1zkNWxAeBQAHjMEGuwZVAC0APiiDpqqADo-vfiERcUEASXgcKRAJeANYAGtydS0dPTBDY1NNc0sbOwcnNzQIqAAfDy8slJgDHABXEHgoAEY0SzgkSBkndwAGHwBWKHs7KAAiZrHeFO0rHzscHCMqnOtbe2AAGigwACMAK29qPHLPEEL3ADkwWOTqzQwggIFhKGucAAtqGgB1cFofqhPgAVL6CGQhcKRaQxOKJNRVXT6IwUFYWNb5RwudxnO73EC1BpNMbAMBgKYpRFzBZLEBo3LrBzbPaHYonCrnbFvG4GPHVR7CXirYAyWDeEJRNS%2BfxEVY0UXed5fAH-X5Aj6g1CCBEpEWwZREIA

Author: Good point! This example is arguably fixable with an `@final`.

I think this doesn't impact the broader point that we can't narrow the attribute type to int, but I think I made a few claims that are too strong when I said these classes don't overlap at all.

```

Indeed, even without such a pathological implementation, since the `x` attribute could be mutable
on implementations of `HasIntStr`,

> the `x` attribute could be modified at any time between an `int` and a `str`. Thus, arguably

Reviewer: "any time between an `int` and a `str`" is an interesting turn of phrase -- I'm not sure when `int` o'clock is ;)

the `HasIntStr` and `HasStr` protocols don't overlap at all and the type of `obj` is equivalent
to just `HasIntStr`.

Are there any variations of the `HasIntStr` + `HasStr` intersection where we *can* narrow the
attribute type? It seems to me that we can only narrow if every value that is a member of
the `HasIntStr` equivalent type must have either an `int` or a `str` attribute, not both.

That is the case, for example, with the following classes:

```python
from typing import NamedTuple, Final, final
from dataclasses import dataclass

class HasIntStrNamedTuple(NamedTuple):
    x: int | str

class HasIntStrFinal:
    x: Final[int | str]

    def __init__(self, x: int | str) -> None:
        self.x = x

@final
@dataclass(frozen=True)
class HasIntStrFrozen:
    x: int | str
```

There might be room for debate over how many of the finality declarations we need
to rule out subclasses that might override the attribute with a descriptor; type checkers

> differ in their behavior in this area. However, the `NamedTuple` case at least seems undisputable.

Reviewer: ```suggestion
differ in their behavior in this area. However, the `NamedTuple` case at least seems indisputable.
```

tracker asking for a `Not` type. Let's review some of the reasons people want this type.

(Note that the original post in the issue actually is *not* asking for negation types, but for
something else. I have an idea for how to support that use case, but I'll talk about that elsewhere.)

#### Type narrowing

The most obvious use case of negation types is in type narrowing, which is very important in practice
in the Python type system. In general, the negative branch of a type narrowing construct (the branch
where the condition does *not* hold) introduces an implicit negation type.
This is explicitly specified for `TypeIs`:

```python
from typing import TypeIs

def is_int(x: object) -> TypeIs[int]:
    # ...

x: int | str
if is_int(x):
    # x is now of type (int | str) & int = int
else:
    # x is of now of type (int | str) & ~int = str
```

Conceptually it's also how most other forms of type narrowing supported by type checkers work.
For example, a check like `x is None` narrows the original type `T` to `T & None` in the positive
case and to `T & ~None` in the negative case; `isinstance(x, int)` narrows to `T & int` in the positive
and `T & ~int` in the negative case.

> In most cases, type checkers handle this through specific special cases; the new `ty` type checker

Reviewer: ```suggestion
In most cases, type checkers handle this through specific special cases; however, the new `ty` type checker
```

soundly. If I change a function to accept only `Sequence[str] & ~str`, any function that accepts a
`Sequence[str]` and passes it on to my function would have to be changed; either I'd have to add
`assert not isinstance(x, str)` (potentially creating a runtime error)
or I'd have to change my own annotation to `Sequence[str] & ~str` too,
pushing the problem on to my callers. This would be especially invasive with the `str & ~LiteralString`
idea; users might ask for virtually all return annotations of `str` in typeshed to be changed
to `str & ~LiteralString`.

However, a nice aspect of these use cases is that they involve fully static types, so we sidestep
the greater complexity of gradual negation types.

#### Accepting all types with some marker type

Another group of use cases involves code that conceptually accepts all objects, but treats
some particular type (often `None`) specially.

An example is this use case by Alexander Martin:

```python
@overload
def foo(a: None) -> str | None:
   ...

@overload
def foo[T: ~None](a: T) -> T:
   ...

def foo(a):
    ...

foo(1)  # Literal[1]
foo(None)  # str | None

> foo(object())  # both overloads match?

Reviewer: shouldn't this be "neither overload matches"? `object` is not assignable to `None`, but it is also not disjoint from `None` (and therefore not assignable to `~None` either)

Oh -- but I suppose that per https://typing.python.org/en/latest/spec/overload.html#argument-type-expansion, `object` could theoretically be "expanded" to `None | ~None`, meaning that the call should not be rejected, and instead we should union the two possible return types together to `((str | None) | ~None) == `object`?

Anyway, you may be interested in https://github.com/astral-sh/ruff/pull/19120, where I removed some overloads like this from ty's implementation of the descriptor protocol because we were selecting the incorrect overload for types that were not disjoint from `None` (https://github.com/astral-sh/ty/issues/737)

Author: Oh yes, this would return the union according to step 3 of the overload algorithm. I'll add a note discussing argument expansion.

Given a value `x: T & ~U`, how should a type checker infer the type of an expression
like `x.attr`? In most cases, once the intersection has been fully simplified, the
negative part of the intersection can simply be ignored. A type like `int & ~Literal[1]`
has the same attributes as plain `int`; `object & ~str` does not differ from `object`.
It is important, of course, that the intersection first gets simplified.
`(int | str) & ~str` does not have the same attributes as `int`.

It gets more complicated with structural types in the negative case. Consider this
program:

```python
from typing import Protocol, ReadOnly

class HasIntStr(Protocol):
    x: ReadOnly[int | str]

class HasStr(Protocol):
    x: ReadOnly[str]

obj: HasIntStr & ~HasStr
reveal_type(obj.x)
```

(Assuming that `ReadOnly` is accepted on protocol attributes, as proposed by
[PEP 767](https://peps.python.org/pep-0767).)

Naively, we might think that this should reveal `int`: `obj.x` looks like it should be
`(int | str) & ~str = int`. But there could be instances of `HasIntStr` that are not
subtypes of `HasStr`, yet might return either an `int` or a `str` when the attribute is
accessed.

> ```python

Reviewer: ````suggestion
```python
from typing import final

````

value of negation types.

## A brief tour of the set-theoretic gradual type system

For the following discussion of negation types to make sense, we first have to review
the core structure of the type system, as outlined in more detail in
["Type system concepts"](https://typing.python.org/en/latest/spec/concepts.html).

Python's type system is conceptually based on set-theoretic types: a type is understood
as a set of possible values. For example, the type `int` includes all values that are
instance of the class `int` or a subclass at runtime, and the type `Literal[43]`
includes only instances of exactly `int` (not a subclass) with the value `43`. Subtyping
is defined through the subset relation: because all members of `Literal[43]` are also
members of `int`, `Literal[43]` is a subtype of `int` and a value of type `Literal[43]`
can be used when an `int` is expected.

But what we just said only applies to what we call _fully static types_, which represent
known sets of possible values. In addition, we have _dynamic types_, which include a
_gradual form_ such as `Any`. (This term does not appear in the spec, but it is useful

> to be able to talk about dynamic types, and the term seems natural.) We use the term

Reviewer 2: My only hesitance about this use of "dynamic type" is that it is common in the literature for the term "dynamic type" to refer to _the_ "fully dynamic" type, that is, `Any`.

But I agree that we need a less contorted term for "non-fully-static types".

_gradual types_ for all types, dynamic or fully static. A dynamic type does not refer to
a single set of possible values, but can refer to any type within a range. We call these
types _materializations_, and in general, operations on gradual types are acceptable if

> there is any materialization of the gradual type that could make the operation work. The

Reviewer 2: Are there "operations" that don't desugar to function/method calls?

It feels a bit imprecise / poorly-defined to me to discuss "operations" and "make the operation work", when I think we could instead rely on desugaring to function/method calls, and reduce everything to a definition of assignability.

Author: This is meant to be an informal description, I don't think it's out of place in that sense?

> most prominent gradual type is `Any`, which can materialize to any type.

Reviewer 2: Do you mean "the most prominent dynamic type" here?

For example, type checkers emit an error if they see an attempt to add an `int` to a
`str` (`int + str`). But `int + Any` is allowed: the `Any` could materialize to a type
that would make the `+` operation work, such as `int` or `float`. Some errors may still
be detected even if `Any` is used, though. Consider the operation `int + (str | Any)`,
adding an `int` to something that may be either a `str` or `Any`. Now, no matter what we
materialize the `Any` to (e.g., `int + str | int`, `int + str | str`,
`int + str | None`), type checkers will reject the operation, because we might be adding
an `int` and a `str`.

If types are sets, we can use set operations on them. The union operation, written as
`|`, has been part of the Python type system since the beginning. The intersection
operation, which is not currently an explicit part of the type system, is conventionally
written as `&`; for example, `int & str` denotes the set of values that are members of
both the `int` and `str` types (which happens to be the empty set). In this article, we

> focus on a third operatoin: negation.

Reviewer 2: Can't believe I found a typo _after_ an @AlexWaygood review.

```suggestion
focus on a third operation: negation.
```

in the spec; I haven't proposed a fix yet since I need to think more about the right
replacement wording, but a better definition of materialiation would allow for the union
or intersection of two materializations to also be a materialization.

Second, and more seriously, the bottom materialization of many dynamic types will be
equivalent to `Never`. An object that is both a `list[int]` and a `list[str]`, for
example, cannot exist. This in turn means that under the spec, all of these dynamic
types are assignable to each other, since they have a shared materialization. This I
believe is another defect in the spec, but we need to think more about how to fix it.
For our purposes in this article, we can simply stick our heads in the sand and pretend
that `Never` and `Bottom[T]` are different types. It's not a very satisfying solution,
but it allows some interesting reasoning. I hope to write more about this topic later.

### Simplifying negation types

Now, with `Top[T]` available as a tool, we can formulate a rule: for a dynamic type `T`,
`~T` is equivalent to `~Top[T] | T`. This is the union of all objects that are not part

> of any materialization of `T`, plus some unknown part of the dynamic type `T`.

Reviewer 2: This is a cool insight!

We can apply this rule to some example types:

- `~Any` is equivalent to `~Top[Any] | Any`. `Top[Any]` is `object`, so this reduces to
  `~object | Any = Never | Any = Any`.
- `~tuple[Any] = ~Top[tuple[Any]] | tuple[Any] = ~tuple[object] | tuple[Any]`.

Whether this counts as simplifying is debatable, since it usually makes for a more
complex type. Type checkers may have a harder time reasoning about this type; for
example, `Sequence[object]` should be assignable to `~tuple[object] | tuple[Any]`
(because the latter can materialize to `~tuple[object] | tuple[object] = object`), but
assignability to unions is usually implemented by checking each arm of the union
separately, which would yield an incorrect result in this case. However, this identity
does mean that if we reason about negation types, we can reason about fully static
negation types alone without loss of generality.

There are a few other obvious rules that can help simplify negation types:

- `~~T = T`
- `~(T | U) = ~T & ~U`
- `~(T & U) = ~T | ~U`

> - `(T | U) & ~V = (T & ~V) | (U & ~V)`

Reviewer 2: This one seems slightly odd to include here, as it's simply distribution of an intersection over a union, it has nothing in particular to do with negations (`~V` could equally well be replaced with `V` here.) But I guess it is still true that it is "a rule that can help simplify negation types" -- it's a rule that can help simplify many intersection types, whether they involve negations or not.

Author: True, I'll drop this one.

accessed.

```python
from typing import final

@final
class HasIntStrSneaky:
    @property
    def x(self) -> int | str:
        return 1 if random() > 0.5 else "1"

    @x.setter
    def x(self, obj: int | str) -> None:
        pass
```

Indeed, even without such a pathological implementation, since the `x` attribute could
be mutable on implementations of `HasIntStr`, the type of the `x` attribute could change
at any time from an `int` to a `str` and vice versa.

Are there any variations of the `HasIntStr` + `HasStr` intersection where we _can_
narrow the attribute type? It seems to me that we can only narrow if every value that is
a member of the `HasIntStr` equivalent type must have either an `int` or a `str`
attribute, not both.

That is the case, for example, with the following classes:

```python
from typing import NamedTuple, Final, final
from dataclasses import dataclass

class HasIntStrNamedTuple(NamedTuple):
    x: int | str

class HasIntStrFinal:
    x: Final[int | str]

    def __init__(self, x: int | str) -> None:
        self.x = x

@final
@dataclass(frozen=True)
class HasIntStrFrozen:
    x: int | str

> ```

Reviewer: am I correct in saying then that even for these examples, the narrowing of the attribute type is only possible here because `int` and `str` are disjoint: an object can inhabit one or the other type, but never both simultaneously?

Author: I don't think that matters. We can narrow to `A | B & ~B` only if the conditions listed further down hold; whether that simplifies to `A` or not is sort of an independent question.

There might be room for debate over how many of the finality declarations we need to
rule out subclasses that might override the attribute; type checkers differ in their
behavior in this area. However, the `NamedTuple` case at least seems indisputable.

Attribute access on a fully simplified intersection with a negated type (`T & ~U`),
then, should differ in behavior from attribute access on the positive part of the
intersection (`T`) if something like the following conditions hold:

- `U` is a Protocol containing either only the attribute under consideration, or also
  some attributes that are present with the right type on all members of `T`. (In the
  latter case, the intersection could be simplified to an equivalent one where the
  Protocol contains only the relevant attribute.)
- Every member of `T` either has an attribute of the type declared by `U`, or an
  attribute of some other type distinct from `U`; they cannot have an attribute that may

>   be of either type.

Reviewer 2: This is pretty unclear to me -- does "distinct" effectively mean "disjunct" here? Or does it mean "not equivalent"? Or does it mean "not a super-type of"?

The "may be of either type" sounds like it's trying to talk specifically about union types, but I think the concept here should be expressible in terms of core set-theoretic type concepts, without treating union types specially.

Author: Yes, this is a bit vague. I wrote a new version that I think is more rigorous.

---

We can formalize this rule as follows. Given a fully simplified intersection
`T & ~P`, where `T` has an
attribute `attr` of type `AT` and `P` has an attribute `attr` of type `AP`,
accessing the attribute `attr` on `T & ~P` should yield type `AT`, unless these
conditions hold:

- `P` is a Protocol containing either only `attr`, or also
  some attributes that are present with the right type on all members of `T`. (In the
  latter case, the intersection could be simplified to an equivalent one where the
  Protocol contains only the relevant attribute.)
- Every object that is a member of `T` is either a member of `P` (that is, it has an
  attribute of the specified type) or a member of a Protocol containing an attribute
  `attr` of type `~AT`. In other words, members of `T` can be divided into objects that
  have an attribute of type `AT` and ones that have an attribute of some other type;
  there cannot be any with an attribute that may return either type.

For example, in the `HasIntStrNamedTuple` example, every instance of the class
either has an `x` attribute of type `int` or an attribute of type `~int` (i.e., `str`, a subtype
of `~int`). Therefore, given a value of type `HasIntStrNamedTuple & ~HasInt`, we can
infer the type as `(int | str) & ~int = str`. But in the earlier case of `HasIntStrSneaky`,
instances of the class are not divisible into two such buckets, so we cannot narrow
the type of the attribute given a type `HasIntStrSneaky & ~HasInt`.

> Additional slight complications arise if the `U` Protocol declares a mutable attribute,

Reviewer: nit: I'd find it easier to follow if you called the protocol type `P` 😆

we mostly have to worry only about assignability _from_ these negation types, not about
assignability _to_ negation types. That's a "mostly" because assignability from negation
types does come up in cases involving contravariance.

For example, consider this program:

```python
from typing import Generic, TypeVar, Sequence

T_contra = TypeVar("T_contra", contravariant=True)

class X(Generic[T_contra]):
    def __init__(self, x: T_contra) -> None: pass
    def send(self, x: T_contra) -> None: pass

def f(x: X[Sequence[str]]): pass

def doit(o: Sequence[str]):
    if not isinstance(o, str):
        x = X(o)  # X[Sequence[str] & ~str]
        f(x)  # should be rejected: X[Sequence[str] & ~str] is not a subtype of X[Sequence[str]]
```

If we reason from basic principles of negation types, this program should be rejected,
and indeed `ty` rejects it. Other type checkers allow it, however, and it's not obvious
that allowing this program would cause any practical issues.

There are other cases where mypy and pyright currently show the wrong result as compared
with the answer expected from our theoretical treatment of negation types:

```python
from typing import TypeIs, Any, reveal_type

> def is_list(x: object) -> TypeIs[list[Any]]:

Reviewer 2: Assuming the intended implementation is `return isinstance(x, list)`, and the mypy/pyright behavior below is the desired narrowing behavior (which seem like likely presumptions in this example), the correct return type here would be `TypeIs[Top[list[Any]]]`, if we had a way to express that (infinite union) type.

Author: I didn't mean for this to just be `isinstance(..., list)`; that has simpler behavior. Maybe a fun example implementation would be `is_exactly_list`: `type(x) is list`.

> the mypy/pyright behavior below is the desired narrowing behavior

Not sure about that. It's not the theoretically correct behavior, and it can lead to unsoundness (I should add some examples to my repo). It might be the practically useful choice though, since otherwise there's no way to represent the behavior of `isinstance(..., list)` using TypeIs.

Reviewer 2: I agree that the pyright/mypy narrowing behavior is not theoretically correct, and can be unsound, for the return type `TypeIs[list[Any]]`. My point was that I think the correct narrowing type for `isinstance(x, list)` should be `TypeIs[Top[list[Any]]` instead, and given that, the mypy/pyright narrowing behavior would be correct.

Author: Sure, we're not using `isinstance(x, list)` here though; the text here is talking about interactions with dynamic types, and `Top[list[Any]]` is fully static. I'll see if I need to rewrite anything here to make it clearer, but hopefully it's enough to use something my `is_exactly_list()` example.

Reviewer 2: Yes, I think the key thing to make this clearer would be to find a realistic example where it is actually correct/sensible to return a non-fully-static `TypeIs` type. I'm not sure that `is_exactly_list` qualifies, but I also don't have other ideas.

(I do think it's a problem that we have no way to spell the correct `TypeIs` type for `isinstance(x, list)`, but I recognize that problem is not what this section is really about.)

Author: Revising this to use `is_exactly_list` and give a subclass of list in the example. It's maybe not a realistic use case but it does demonstrate the unsoundness.

>     raise NotImplementedError

Reviewer 2: I'm curious why you use `raise NotImplementedError` here, rather than an actual implementation.

def f1(x: list[Any] | str):
    if is_list(x):
        reveal_type(x)  # list[Any]
    else:

>         reveal_type(x)  # mypy and pyright: str (incorrect, should be str | list[Any])

Reviewer 2: Amusingly, ty says `(list[Any] & ~list[Any]) | str` here, but I think you're right that ideally we would simplify `(list[Any] & ~list[Any])` to just `list[Any]`.

Author: It's not wrong! I don't think you necessarily need to simplify everything.

    def get(self, key: ~KT, /) -> None: ...
    @overload
    def get(self, key: object, /) -> VT_co | None: ...
    @overload
    def get[T](self, key: ~KT, /, default: T) -> T: ...
    @overload
    def get[T](self, key: object, /, default: T) -> VT_co | T: ...

m: Mapping[str, int]
m.get(1)  # None
m.get(1, 2)  # Literal[2]
m.get(object())  # int | None
m.get(object(), "default")  # int | Literal["default"]
```

These use cases don't have as many issues with virality as the previous category, but
they are more likely to require advanced aspects of negation types.

### Pragmatic negation types

Above, we came up with one category of use cases (type narrowing) where type checkers
must have some notion of negation types, but only implicit negation types and only in a
somewhat limited context. There are also some use cases that would require explicit
negation types, but those are extensions of the current type system. We also saw above
that full support for gradual negation types requires significant amounts of machinery
that the current type system does not require, such as the top and bottom
materialization; in addition, it becomes very important that type checkers can correctly

> check whether or not an intersection is inhabited.

Reviewer 2: I'm not sure that top and bottom materialization are strictly required to support gradual negation types; it depends on the implementation strategy. In the case of ty, "correctly check whether or not an intersection is inhabited" is implemented directly/manually as a fundamental type relation (disjointness).

Author: There might be a way but ty's behavior is incorrect in a number of cases. In particular, the assignability rule I formulate for assignability between gradual negation types involving checking the disjunctness of the bottom materializations of both. That's different from whether the types themselves are disjunct.

For example, ty issues an error for this:

```python
from typing_extensions import Any
from ty_extensions import Not

def f(x: tuple[str | Any]):
    g(x)

def g(x: Not[tuple[int | Any]]):
    pass
```

That's incorrect; if we materialize both instances of Any to Never, we get `tuple[str]` and `Not[tuple[int]]`, and the former is assignable to the latter. If ty rejects this, it may be checking whether `tuple[str | Any]` and `tuple[int | Any]` are disjunct, which is the wrong thing to check.

I'll hedge this a bit because maybe there are implementation strategies that don't rely on the bottom materialization.

Reviewer 2: Thanks; I only today reached https://github.com/astral-sh/ty/issues/767 in working through my notifications from vacation :) I think your analysis of what ty is doing wrong here is exactly correct, and it does seem like bottom materialization is the way to fix it.
